### PR TITLE
[MRG] DOC: Add new changelog template to contributors' guide

### LINKED ISSUE #8436: "If comment is empty, a 1-based index is assigned by Richard Höchenberger" – What's New entries
Hello, I just wanted to inquire if we could, maybe, agree to always add a comma before the contributors' names in our changelogs? Otherwise we keep producing sometimes confusing, sometimes rather hilarious lines like the one in the title.

The latest changelog(s) have been rather inconsistent in that manner, although my impression is that more and more of those commas have made it through the review process.

So what I'm proposing is that instead of:

> If comment is empty, a 1-based index is assigned by Richard Höchenberger

we'd add an entry

> If comment is empty, a 1-based index is assigned, by Richard Höchenberger

WDYT?

cc @cbrnr @agramfort @sappelhoff @larsoner @drammock

### PR DESCRIPTION
**Pull request by hoechenberger**
Also touch two changelog entries to see what it looks like when rendered.

Closes #8436

### CODE REVIEW

**Comment by hoechenberger** [comment]
Updated guide: https://23036-1301584-gh.circle-artifacts.com/0/dev/install/contributing.html#describe-your-changes-in-the-changelog

Example of updated changelog entries:

**Comment by agramfort** [comment]
see how sklearn does it:
https://scikit-learn.org/stable/whats_new/v0.23.html#id5

it's:

 #PRNB by Author.

>

**Comment by cbrnr** [comment]
We could put that in parens, i.e.

- When reading BrainVision raw data, the channel units and types were sometimes not inferred correctly (#8434 by Richard Höchenberger)

**Comment by hoechenberger** [comment]
Looking good

**Comment by drammock** [comment]
The revised order suggested by @agramfort and @cbrnr is fine with me. While you're editing that section of the docs, please change the part about `doc/whats_new.rst` to `doc/changes/latest.rst`

**Comment by larsoner** [comment]
Agreed, let's go with that. It's nice to stick to something already used by another (more popular) library, too.

**Comment by hoechenberger** [comment]
Artifact:
https://23060-1301584-gh.circle-artifacts.com/0/dev/install/contributing.html#describe-your-changes-in-the-changelog

Updated part of the changelog:
Screenshot 2020-10-29 at 09 36 11

**Comment by hoechenberger** [comment]
I've also addressed @drammock's comment.

**Comment by hoechenberger** [comment]
Umm, the link to the "changelog" at https://23060-1301584-gh.circle-artifacts.com/0/dev/install/contributing.html#describe-your-changes-in-the-changelog leads to scikit-learn.org, I thought this was only a problem when I built the docs locally, but it's present here too. Needs to be fixed, but I don't know how.

**Comment by cbrnr** [comment]
Same here: https://github.com/mne-tools/mne-python/blob/0b7052f25de10c5f8a825439ae5f7a7805d3df51/doc/install/contributing.rst#L855

**Comment by hoechenberger** [comment]
Yes, and it's in the stable docs too

**Comment by drammock** [comment]
The first thing I would try is changing `:doc:` to `:mne:doc:` to see if it resolves the name conflict locally

**Comment by larsoner** [comment]
I *think* to specify the ref it would need to be inside the backticks/link like
```
:doc:`changelog <mne:whats_new>` 
```
or so. I would do a local `make html_dev-noplot` to check.

But really I think it probably just needs to be:
```
:ref:`changelog <whats_new>`
```

**Comment by hoechenberger** [comment]
@drammock 
> The first thing I would try is changing `:doc:` to `:mne:doc:` to see if it resolves the name conflict locally
@larsoner 


> ```
> :doc:`changelog <mne:whats_new>` 
> ```

neither worked. But 
@larsoner 


> ```
> :ref:`changelog <whats_new>`
> ```

this produced the desired result. Pushed this change.

File: doc/install/contributing.rst [hunk 1]
```diff
@@ -538,12 +538,31 @@ Describe your changes in the changelog
...
+
+.. code-block::
+
+    - Short description of the changes from one contributor in multiple PRs (:gh:`0000`, :gh:`1111` by `Contributor Name`_)
+    - Short description of the changes from several contributors in multiple PRs (:gh:`0000`, :gh:`1111` by `Contributor Name`_, `Second Contributor`_, and `Third Contributor`_)
+
+Note that there are sections of the changelog for each release, and separate
+subsections for bugfixes, new features, and changes to the public API. It is
+usually best to wait to add a line to the changelog until your PR is finalized,
+to avoid merge conflicts (since the changelog is updated with almost every PR).
```
Comment (larsoner) [inline doc/install/contributing.rst:565]: Can you add a note about how new contributors should put their entry at the top of the appropriate section (with an example with the RST substitution business) and add their name to `names.inc`

**Comment by cbrnr** [comment]
@hoechenberger if you have time it would be good to get this merged, because new changes keep coming in 😄!

**Comment by hoechenberger** [comment]
@cbrnr This one totally slipped my mind, sorry! Will have a look asap.

**Comment by cbrnr** [comment]
No worries, let me know when it is ready to review.

**Comment by hoechenberger** [comment]
@cbrnr I think I'm done here

**Comment by hoechenberger** [comment]
(just rebased)

File: doc/install/contributing.rst (hunk 1 above)
Comment (hoechenberger) [reply doc/install/contributing.rst:565]: Done

Comment (cbrnr) [reply doc/install/contributing.rst:565]: You didn't add the part about adding to the top of the section. @larsoner is this really how we handled it? I always put new entries at the bottom of the appropriate section. In any case, this should be mentioned here.

**Review by cbrnr — APPROVED** [review]
Other than my comment about the changelog entry 👍!

File: doc/install/contributing.rst (hunk 1 above)
Comment (hoechenberger) [reply doc/install/contributing.rst:565]: At the top now?? Why's that?

Comment (cbrnr) [reply doc/install/contributing.rst:565]: I think @larsoner meant at the bottom of the appropriate section.

Comment (cbrnr) [reply doc/install/contributing.rst:565]: But we should make sure to mention this (whatever we decide, probably at the bottom).

**Comment by hoechenberger** [comment]
@cbrnr I've tried to address this, can you have another look please? I can also ping you again once the doc has been rendered.

**Comment by hoechenberger** [comment]
@cbrnr Built docs: https://23342-1301584-gh.circle-artifacts.com/0/dev/install/contributing.html#describe-your-changes-in-the-changelog

File: doc/install/contributing.rst (hunk 1 above)
Comment (larsoner) [reply doc/install/contributing.rst]: > You didn't add the part about adding to the top of the section. @larsoner is this really how we handled it? I always put new entries at the bottom of the appropriate section. In any case, this should be mentioned here.

Above I'm talking about *new contributors* -- we decided their entries should go at the top of the appropriate section. This should be added along with an example with the substitution syntax that allows for bolding of their name within a link as well.

Comment (hoechenberger) [reply doc/install/contributing.rst]: Ooooh!! Ok that does make sense indeed

**Comment by cbrnr** [comment]
Looks good! @larsoner please go ahead and merge if you're happy!

**Comment by hoechenberger** [comment]
@cbrnr He's not yet happy :) I'll need to address https://github.com/mne-tools/mne-python/pull/8442#discussion_r519768069

Will do after a short walk outside

**Comment by cbrnr** [comment]
Ah, didn't see that comment!

**Comment by hoechenberger** [comment]
So it seems I cannot use
```
Foo
^^^
```
to create a sub-section nested below
```
Bar
---
```
in our docs. Which underline characters would I need to use instead?

**Comment by drammock** [comment]
@hoechenberger Sphinx computes character-to-heading-level mappings on a per-document basis.  In that particular document, `^^^` has already been used as a second-level heading (after `===` for the title), so you can't also use it below `---` (which is a fourth-level heading in that document, nested under 3rd-level `~~~`).

As for what is legal, here is [the sphinx doc section](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections) and here is [the relevant section for rST](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#sections)

**Comment by hoechenberger** [comment]
@drammock Wow, this is complicated… thanks for the explanation!

**Comment by hoechenberger** [comment]
Doc build succeeds now!

https://23389-1301584-gh.circle-artifacts.com/0/dev/install/contributing.html#describe-your-changes-in-the-changelog

Note that I've also changed the TOC to show one additional level of nesting:
https://23389-1301584-gh.circle-artifacts.com/0/dev/install/contributing.html

File: doc/install/contributing.rst [hunk 2]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
-:doc:`changelog <whats_new>` (:file:`doc/whats_new.rst`; this can be skipped
-for very minor changes like correcting typos in the documentation). Note that
-there are sections of the changelog for each release, and separate subsections
-for bugfixes, new features, and changes to the public API. It is usually best
-to wait to add a line to the changelog until your PR is finalized, to avoid
-merge conflicts (since the changelog is updated with almost every PR).
+:ref:`changelog <whats_new>` (:file:`doc/changes/latest.inc`; this can be
+skipped for very minor changes like correcting typos in the documentation).
 
+.. note::
```
Comment (larsoner) [inline doc/install/contributing.rst:544]: `.. note::` has a tendency to break up the flow. I would not make this a `note` but rather just a standard paragraph.

File: doc/install/contributing.rst [hunk 3]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
+"""""""""""""""""""""""
+
+Welcome to MNE-Python! We're very happy to have you here. 🤗 And to ensure you
+get proper credit for your work, please add a changelog entry with the
+following pattern **at the top** of the respective subsection (bugfix,
+new feature etc.):
+
+.. code-block::
+
+  - Short description of the changes **by new contributor** |Your Name|_ (:gh:`0000`)
```
Comment (larsoner) [inline doc/install/contributing.rst:562]:
```suggestion

  Bug
  ---

  .. |Your Name| replace:: **Your Name**

  - Short description of the changes (:gh:`0000` **by new contributor** |Your Name|_)
  
  - ...
```

File: doc/install/contributing.rst [hunk 4]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
+
+First-time contributors
+"""""""""""""""""""""""
+
+Welcome to MNE-Python! We're very happy to have you here. 🤗 And to ensure you
+get proper credit for your work, please add a changelog entry with the
+following pattern **at the top** of the respective subsection (bugfix,
+new feature etc.):
+
+.. code-block::
```
Comment (larsoner) [inline doc/install/contributing.rst:560]:
```suggestion
.. code-block:: rst
```

File: doc/install/contributing.rst [hunk 5]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
+
+  - Short description of the changes **by new contributor** |Your Name|_ (:gh:`0000`)
+
+where ``0000`` must be replaced with the respective GitHub pull request (PR)
+number.
+
+.. note::
```
Comment (larsoner) [inline doc/install/contributing.rst:567]: I wouldn't make this a `note` either

File: doc/install/contributing.rst [hunk 6]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
+
+Many contributors opt to link to their GitHub profile that way. Have a look
+at the existing entries in the file to get some inspiration.
+
+Recurring contributors
+""""""""""""""""""""""
+
+The chngelog entry should follow the following patterns:
+
+.. code-block::
```
Comment (larsoner) [inline doc/install/contributing.rst:589]:
```suggestion
.. code-block:: rst
```

File: doc/install/contributing.rst [hunk 7]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
+  It is usually best to wait to add a line to the changelog until your PR is
+  finalized, to avoid merge conflicts (since the changelog is updated with
+  almost every PR).
+
+Lastly, make sure that your name is included in the list of authors in
+:file:`doc/changes/names.inc`, otherwise the documentation build will fail.
+To add an author name, simply append a line with the following pattern (note
+how the syntax is different from that used in the changelog):
+
+.. code-block::
```
Comment (larsoner) [inline doc/install/contributing.rst:577]:
```suggestion
.. code-block:: rst
```

File: doc/install/contributing.rst [hunk 8]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
+
+  .. _Your Name: https://www.your-website.com/
+
+Many contributors opt to link to their GitHub profile that way. Have a look
+at the existing entries in the file to get some inspiration.
+
+Recurring contributors
+""""""""""""""""""""""
+
+The chngelog entry should follow the following patterns:
```
Comment (larsoner) [inline doc/install/contributing.rst:587]:
```suggestion
The changelog entry should follow the following patterns:
```

File: doc/install/contributing.rst [hunk 9]
```diff
@@ -538,13 +538,70 @@ Describe your changes in the changelog
...
+    - Short description of the changes from several contributors (:gh:`0000` by `Contributor Name`_, `Second Contributor`_, and `Third Contributor`_)
+
+where ``0000`` must be replaced with the respective GitHub pull request (PR)
+number. Mind the Oxford comma in the case of multiple contributors.
+
+Sometimes, changes that shall appear as a single changelog entry are spread out
+across multiple PRs. In this case, simply name all relevant PRs, separated by
+commas:
+
+.. code-block::
```
Comment (larsoner) [inline doc/install/contributing.rst:601]:
```suggestion
.. code-block:: rst
```

File: doc/changes/latest.inc [hunk 10]
```diff
@@ -99,9 +99,9 @@ Bugs
 
 - Fix bug in :func:`mne.set_bipolar_reference` where ``ch_info`` could contain invalid channel information keys by `Eric Larson`_
 
-- When reading BrainVision raw data, the channel units and types were sometimes not inferred correctly by `Richard Höchenberger`_ (:gh:`8434`)
+- When reading BrainVision raw data, the channel units and types were sometimes not inferred correctly (:gh:`8434` by `Richard Höchenberger`_)
 
-- Attempting to remove baseline correction from preloaded `~mne.Epochs` will now raise an exception by `Richard Höchenberger`_ (:gh:`8435`)
+- Attempting to remove baseline correction from preloaded `~mne.Epochs` will now raise an exception (:gh:`8435` by `Richard Höchenberger`_)
```
Comment (larsoner) [inline doc/changes/latest.inc:104]: We should probably change all in `latest.inc` in this PR

**Comment by hoechenberger** [comment]
Addressed all comments.

**Comment by larsoner** [comment]
Thanhks @hoechenberger !

**Pull request merged**